### PR TITLE
Use EDA User Service to persist analysis configs and preferences

### LINKED ISSUE #441: General: Wire up the user service for analyses
When the EdaUserService has been completed, we will want to update the client code to utilize this new service. The service will be used to retrieve analysis summaries for all listing pages, and analysis objects for the workspace component. It will also be used to retrieve and persist preferences, such as pinned analyses. 

- [x] Add a configuration for `EDA_USER_SERVICE_URL`. This value should be used in `setupProxy.js` to create a proxy to the service. Follow the same pattern used for `EDA_SUBSETTING_SERVICE_URL` and `EDA_DATA_SERVICE_URL`.
- [x] Ensure the API is satisfied by the current [analysis api adaptor](https://github.com/VEuPathDB/web-eda/blob/1dada641d3353fb21a4a0b090f456af6c9318549/src/lib/core/api/analysis-api.ts). Changes will likely be required.
- [x] Update the `mockAnalysisStore` to satisfy the new analysis adaptor interface.
- [x] Fix all TypeScript compilation errors which arise. In particular, changes may need to be made to the various hooks in [src/lib/core/hooks/analysis.ts](https://github.com/VEuPathDB/web-eda/blob/4ef6d36ec57b8bdff7c38821f3cf29e93eb95024/src/lib/core/hooks/analysis.ts).
- [x] Add a new prop to `WorkspaceRouter` named `userServiceUrl` and use this prop to initialize an `AnalysisClient` instance.

Consult @dmfalke with any questions.

### PR DESCRIPTION
**Pull request by jtlong3rd**
Resolves #441.

Broadly speaking, this PR updates the `AnalysisClient` interface to conform to the [EDA User Service API](https://veupathdb.github.io/EdaUserService/api.html), and introduces an `AnalysisClient` implementation which exercises that service.

**Notes**:
1. The "footprint" of this PR is deceptively large - most of changes here are due to changes in the structure of the "analysis" JSON. Most of the differences in structure are straightforward (e.g., `analysis.starredVariables` is now `analysis.descriptor.starredVariables`), but two nontrivial changes are worth noting here:
   - `StarredVariables` is now an array of `VariableDescriptor`s, as opposed to an array of (string) variable IDs.
   - Computations and visualizations are now stored in a nested JSON structure, as opposed to two "flat" arrays of computations and visualizations. Now, each analysis has an array of `Computation`s, and each `Computation` has an array of `Visualization`s.
2. I have updated the mock analysis client in case we want to use it in the future. (In particular, since we don't yet have WDK login for the local dev environment, it might be worth adding some sort of `USE_MOCK_ANALYSIS_STORE` environment variable.)

**Future work**:
1. I uncovered a preexisting bug with copying analyses - after the copy is complete, the client redirects to the copy's URL, but the page's analysis state is not changed. I'm guessing we just need to make a small update to the [analysis hook](https://github.com/VEuPathDB/web-eda/blob/main/src/lib/core/hooks/analysis.ts).
2. It appears that EDA services on our servers (`ash`, `fir`, etc.) are sending responses with a `set-cookie` header that resets the `JSESSIONID` used by the WDK service. Maybe we need to add some sort of (backend) filter to prevent this from happening?
3. Fix the client unit tests. (Some of them are [and were previously] broken due to what looks like a transpilation error.)

### CODE REVIEW

**Review by dmfalke — COMMENTED** [review]
This looks really great! Before I approve this, I would like to make a suggestion, pertaining to src/lib/core/hooks/analysisClient.ts and how the `wdk_check_auth` cookie is obtained.

The way it is currently done means an `AnalysisClient` instance is not immediately available. As you know, this leads to a number of `undefined` checks. I think part of the motivation to do things the way you have has to do with obtaining the `WdkService` instance needed to get the cookie value. It's most convenient to use the `useWdkService` hook to do this.

I'm wondering if we could instead think of a way to integrate this acquisition of the cookie into each `AnalysisClient` method that needs it. One way I can think of doing this is to pass an instance of `WdkService` to the constructor of `AnalysisClient`. Then, each method that needs the cookie can use a private, memoized method that returns the cookie value. This is just one possibility... there may be others.

Let me know what you think!

**Comment by bobular** [comment]
Hi @jtlong3rd  or @dmfalke  

How would I test this from my developer instance?

I have tried
```
EDA_USER_SERVICE_URL=https://qa.clinepidb.org/eda-user-service
```
and
```
EDA_USER_SERVICE_URL=https://qa.clinepidb.org/eda-data
```
in .env.local

I'm getting 404s from any eda-user-service requests

**Comment by jtlong3rd** [comment]
> Hi @jtlong3rd or @dmfalke
> 
> How would I test this from my developer instance?
> 
> I have tried
> 
> ```
> EDA_USER_SERVICE_URL=https://qa.clinepidb.org/eda-user-service
> ```
> 
> and
> 
> ```
> EDA_USER_SERVICE_URL=https://qa.clinepidb.org/eda-data
> ```
> 
> in .env.local
> 
> I'm getting 404s from any eda-user-service requests

Hi @bobular, if you're testing right now, I would use `EDA_USER_SERVICE_URL=https://qa.clinepidb.org/eda-data/users`.

Per a good suggestion of @dmfalke's above (re: handling the "public analyses" endpoint), I will be changing the expected config for `EDA_USER_SERVICE_URL`. (Working on that right now.) I'll leave a comment here when that's in place, so that you know when to change your config back.

File: src/lib/core/api/analysis-api.ts [hunk 1]
```diff
@@ -1,11 +1,40 @@
...
+
+export type SingleAnalysisPatchRequest = Partial<
+  Pick<Analysis, 'displayName' | 'description' | 'descriptor' | 'isPublic'>
+>;
+
 export class AnalysisClient extends FetchClient {
+  static getClient = memoize(
+    (config: AnalysisClientConfiguration): AnalysisClient =>
+      new AnalysisClient({
+        baseUrl: `${config.userServiceUrl}/${config.userId}`,
```
[the remark this answers is not in the document]
Comment (jtlong3rd) [reply src/lib/core/api/analysis-api.ts]: Updated!

**Comment by bobular** [comment]
Thanks - it's working now - just as a guest/unlogged in user.
For some reason the first analysis I made wasn't saved, but the second two were.  I'll figure out how to reproduce if it happens again.

**Comment by jtlong3rd** [comment]
@bobular - I've updated the PR with @dmfalke's suggested change to the `userServiceUrl` format. You should now be able to use `EDA_USER_SERVICE_URL=https://qa.clinepidb.org/eda-data` in your config!

**Comment by bobular** [comment]
Thanks, this works now.

Is there any intention to persist the guest user ID between react dev server restarts?

Is that what the `WDK_CHECK_AUTH` variable is about?  Could you provide guidance on how to use that please?

**Comment by jtlong3rd** [comment]
> Thanks, this works now.
> 
> Is there any intention to persist the guest user ID between react dev server restarts?
> 
> Is that what the `WDK_CHECK_AUTH` variable is about? Could you provide guidance on how to use that please?

Alas, right now, the guest user is "lost" between react dev server restarts.

As you guessed, the `WDK_CHECK_AUTH` variable is a temporary workaround. Specifying this variable lets you pass as a registered user on the react dev server.

Here's how you can get your hands on a valid `WDK_CHECK_AUTH` value:

1. Go to [ClinEpiDB](https://clinepidb.org/ce/app) (or any WDK site)
2. Log in
3. In your dev tools, find the cookie with key `wdk_check_auth`
4. Use the value of that cookie for your `WDK_CHECK_AUTH` env variable

**Review by dmfalke — APPROVED** [review]
This looks great! I'm happy to go forward with this, but I one other improvement to `AnalysisClient` could be to override the `fetch` method to inject the `wdk_check_auth` header, rather than repeating in each method.

**Comment by bobular** [comment]
Thanks - that works for me now, between server restarts.

I again had the problem where the very first analysis I made **and named** disappeared.  I can't reproduce it though :-( 

I won't be able to check every line of code, but I guess it would be good to merge this sooner rather than later, because of all the potential merge conflicts.  So don't let me stop you!

**Comment by jtlong3rd** [comment]
> This looks great! I'm happy to go forward with this, but I one other improvement to `AnalysisClient` could be to override the `fetch` method to inject the `wdk_check_auth` header, rather than repeating in each method.

This is simple enough to act on now! I'll do that before merging. I also wanted to memoize the fetching of the "user request metadata" (user id, wdk_check_auth) per your original suggestion.

**Pull request merged**